docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md

```diff
@@ -0,0 +1,264 @@
+  select(peptideid, go_id)
+
+pfam <- read_csv("~/Downloads/backup_peptipedia_29_03_2023/table_public.pfam.csv",
+                 col_names = c("peptideid", "pfam_id", "pfam_name", "pfam_type"))
+
+final_dataset <- final_dataset %>%
+  left_join(gene_ontology, by = c("member" = "peptideid")) %>%
+  left_join(pfam, by = c("member" = "peptideid")) %>%
+  distinct()
+
+final_dataset %>% group_by(group_bioactivity) %>% tally()
+
+write_tsv(final_dataset, "~/Downloads/backup_peptipedia_29_03_2023/peptides_for_ml_tmp.tsv")
+```
+
+
+6. Convert the output TSV to a FASTA file that can be used to calculate peptide characteristics (molecular weight, etc).
+```
+cut -f2,3 peptides_for_ml_tmp.tsv | tail -n +2 | seqkit tab2fx > peptides_for_ml.fasta
+```
+
+7. Calculate peptide characteristics (length, mw, etc.)
+
+```
+curl https://raw.githubusercontent.com/Arcadia-Science/peptigate/main/scripts/characterize_peptides.py
+python characterize_peptides.py peptides_for_ml.fasta peptides_for_ml_characterized.tsv
+```
+
+8. Combine peptide metadata with other information
+
+```
+peptide_characterization <- read_tsv("~/Downloads/backup_peptipedia_29_03_2023/peptides_for_ml_characterized.tsv")
+
+final_dataset <- final_dataset %>%
+  left_join(peptide_characterization, by = c("member" = "peptide_id")) %>%
+  mutate(length = nchar(member_sequence)) %>%
+  select(-rep) %>% # remove the mmseqs2 rep column since it shouldn't be used for prediction and is an artifact of dataset creation.
+  rename(peptipedia_peptide_id = member, peptide_sequence = member_sequence)
+
+write_tsv(final_dataset, "peptides_ml_dataset.tsv")
+```
+
+9. Check that the data set is predictive of the label "group_bioactivity".
+```
+library(caret)
+
+set.seed(123)
+
+final_dataset <- read_tsv("petides_ml_dataset.tsv")
+final_dataset_filtered <- final_dataset %>%
+  select(-peptipedia_peptide_id, -peptide_sequence, -all_bioactivities, -go_id, -pfam_id,
+         -pfam_name, -pfam_type)
+trainingIndex <- createDataPartition(final_dataset_filtered$group_bioactivity, p = .8, list = FALSE)
+trainingData <- final_dataset_filtered[trainingIndex, ]
+testData <- final_dataset_filtered[-trainingIndex, ]
+
+control <- trainControl(method="cv", number=10) # 10-fold cross-validation
+model <- train(group_bioactivity ~ ., data=trainingData, method="rf", trControl=control)
+
+predictions <- predict(model, newdata=testData)
+confusionMatrix(predictions, as.factor(testData$group_bioactivity))
+```
+
+which results in:
+```
+Confusion Matrix and Statistics
+
+                                Reference
+Prediction                       antimicrobial coagulation_or_anticoagulation growth_factor hormone immunological_activity metabolic neurological_activity propeptide signal_peptide therapeutic
+  antimicrobial                             67                              0             0       0                      0         0                     0          0              0           0
```

Comment:
this confusion matrix is a little "too good to be true", which either means that the task is "too easy" -- the classes are too easily-separable -- or that there's some unintended leak of the class into the features. for a dataset that is aimed at learning about ML, I wonder if it's better to have a harder task, one where a random forest with no tuning doesn't do better than 0.7-0.8 accuracy. that will allow showcasing the performance of different models, impact of hyperparams, interesting outliers, etc.

Reply:
ok I was able to decrease the out-of-the-box accuracy down to 92% (from 99% I believe) by decreasing the mmseqs cluster threshold from 30% to 10%, and by decreasing the minimum cluster size from 50 to 10. 

This is still not 70-80% accuracy. The things I can think to do to decrease the accuracy further are:
1. create separate training and test data sets explicitly, where sequences that cluster together won't be split between train and test. This should be pretty easy to do since we record cluster representatives, and will decrease the leakage.
2. I'm thresholding pretty stringently to only keep groups where all peptides have at least 1 bioactivity label that is the same for all peptides. I could drop that bar to something like 70% of peptides in a group, and just filter out the ones that don't have it or artificially assign the peptides in that group to have that label


Decreasing the group size to 10 also has the consequence of including less common groups. I think I potentially should also probably remove groups that are smaller than X (50? 200?) and decrease some of the homogeneity of the classes (cluster at 80% and select representatives?).

|group_bioactivity              |    n|
|:------------------------------|----:|
|signal_peptide                 | 3447|
|antimicrobial                  | 2254|
|therapeutic                    | 1352|
|immunological_activity         |  817|
|propeptide                     |  627|
|hormone                        |  421|
|growth_factor                  |  223|
|coagulation_or_anticoagulation |  222|
|neurological_activity          |  220|
|metabolic                      |   93|
|antibacterial_antibiotic       |   90|
|bioactive                      |   59|
|allergen                       |   56|
|toxins                         |   39|
|anti_gram                      |   33|
|amphibian_defense              |   29|
|anti_hiv                       |   27|
|cell_penetrating               |   26|
|defense                        |   26|
|cell_cell_communication        |   22|
|cell_sensing                   |   22|
|anticancer                     |   16|
|other_activity                 |   13|
|angiogenic                     |   12|
|antifungal                     |   12|
|nematode                       |   12|

I've never worked so hard to make a model less accurate 😂

Reply:
Ok I did two things:
1. Removed classes that only have one representative cluster --  all of these were sequences that all clustered together (10% or greater) in the initial clustering, meaning they are all similar. If I had included these, there wouldn't be any way to have data in the train and test sets that were from different clusters, so I just removed them.
2. split the test and training data sets so that sequences from the same clusters all ended up in the training or in the testing, so there wasn't pollution between the two data sets. this dramatically reduces the performance (as one would expect). Model performance summary is now:
   ```
   Overall Statistics
                                          
               Accuracy : 0.5808          
                 95% CI : (0.5566, 0.6047)
       No Information Rate : 0.2756          
       P-Value [Acc > NIR] : < 2.2e-16       
                                          
                     Kappa : 0.4825          
                                          
    Mcnemar's Test P-Value : NA   
   ```

not totally sure how to provide this data -- maybe as separate training and testing csv files?

Reply:
oh and here's the breakdown of the number of clusters (NOT number of sequences) that end up in each type, train and test.
```
                                 test train
  allergen                          1     1
  amphibian_defense                 1     1
  anti_hiv                          1     1
  antibacterial_antibiotic          2     3
  antimicrobial                    32    72
  bioactive                         1     2
  cell_penetrating                  1     1
  cell_sensing                      1     1
  coagulation_or_anticoagulation    2     2
  growth_factor                     3     5
  hormone                           7    14
  immunological_activity           12    25
  metabolic                         1     1
  neurological_activity             3     6
  propeptide                        6    14
  signal_peptide                   38    87
  therapeutic                      19    43
  toxins                            1     1
  ```

Reply:
This makes sense to me! FWIW, I would've cheated and just randomly shuffled a random subset of the rows in each of the feature columns. Obviously this would result in a patently "fake" dataset, but you'd be able to control the extent of the decreased accuracy by how many rows you shuffled. Anyway, I think this is a realistic and realistically hard dataset that will work for the AUG lessons. 

One issue: for some parts of the lessons it will be important to dynamically split the dataset into train and test (e.g. for doing cross-validation), so we can't really have static training and test sets. But I don't think splitting on cluster is necessary - the 58% accuracy feels a bit *too* low to me. As long as they're not huge, I would just commit the final merged dataset to the repo. If you want to retain the cluster-based train/test split, you could add a column containing a flag for this split, so we can recapitulate it in the lessons when we need an extra hard example.

Reply:
Hmmm ok. I could do the "right" thing that I would do if I were actually building a bioactivity classifier and cluster the entire data set at 80%. Then, I could still only select clusters that have consistent bioactivity labels (which would also reduce the size of the data set and increase signal although somewhat artificially), and then see how the classifier does. This will have the side effect of removing pollution between train and test no matter how it is carved up, and you won't need the representative sequence ID to make sense of things. This will be quick for me to try so I'll run it and report back!

Point well taken about shuffling rows -- would definitely be way more efficient. It makes me a little nauseous thinking of putting out fake data like that into the world tho...if I can pull off making this data set with semi-real data, then I'd like that, but if I don't accomplish the thing on this next iteration let's go with your approach 😂

Reply:
weeeeew it worked! I totally didn't think it was going to -- 

```
Overall Statistics
                                          
               Accuracy : 0.6188          
                 95% CI : (0.5805, 0.6559)
    No Information Rate : 0.2496          
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.5316          
                                          
 Mcnemar's Test P-Value : NA   
 ```
 
 and here are the final classes that got included:
 ```
 |bioactivity              |   n|
|:------------------------|---:|
|antimicrobial            | 828|
|signal_peptide           | 783|
|propeptide               | 438|
|therapeutic              | 382|
|immunological_activity   | 307|
|hormone                  | 180|
|neurological_activity    | 111|
|antibacterial_antibiotic |  97|
|amphibian_defense        |  88|
|growth_factor            |  63|
|anti_gram                |  50|
```

Since we removed homogeneity, this data set should be able to be used any which way that anyone wants to. The only thing it can't be used for would be removing homogeneity, but even then you could just drop the threshold from 80% to 70% or 50% and it might still work

Reply:
Nice! This looks like it'll work. And it has the virtue of being entirely authentic (I agree generating a fake dataset doesn't *feel* great, and I hadn't thought about the fact that this will all be public, so the bar should be a bit higher than if we were just doing this for our own internal purposes).

Reply:
+1 this is great. thanks for the modifications!